### ISSUE TITLE
SkSL-based shader warmup

### ISSUE THREAD

**Issue by liyuqian** [flutter/flutter#53607]
This tracks our short-term solution of https://github.com/flutter/flutter/issues/32170 that's based on SkSL (Skia shader language).

- [x] Automatic shader-compile jank removal using SkSL precompile for customer: dream #40686
- [x] Create service protocols to read back generated SkSLs #53114
- [x] Connect flutter tools to VM service callback to pull sksl files from device #53116
- [x] Add support for --sksl-shader-warmup to package sksl files into flutter_assets #53115
- [x] Allow reading SkSLs artifacts from a special asset subdirectory #53117
- [x] Update SkSL bundle format #55219
- [x] Automatically write SkSLs with --cache-sksl flag #54484
- [x] Document how to use SkSL-based shader warmup #60313

Everything above would allow developers to use SkSL shader warm-up from command lines. We will also add support for IDE plugins #54287 in the future.

**Comment by liyuqian** [comment]
I recently did an experiment for SkSL cross-device feasibility regarding the comment in https://github.com/flutter/flutter/pull/53278#issuecomment-605396280. (CC @jonahwilliams @brianosman.) I have not done the cross-Android-iOS experiment yet (I'm planning to do it soon), but the results from two vastly different Android devices make me feel quite optimistic. 

Two Android devices that I tested are
- [2016 low-end Moto G4](https://en.wikipedia.org/wiki/Moto_G4)
- [2019 high-end 90Hz OnePlus 7 pro](https://en.wikipedia.org/wiki/Moto_G4)

After the `flutter_gallery__transition` driver test, both devices generated 35 SkSLs. 33 of them have the same keys (filenames). The content of them have slight differences, but because they have the same keys, the `PersistentCache` would be able to load one cache generated from another device.

I then copied the SkSLs from one device to another device, and see the following results

| Worst frame time (ms)      | MotoG4 | OnePlus7Pro |
|----------------------------|--------|-------------|
| No cache                  | 94.535 | 38.398      |
| Binary cache              | 36.184 | 19.029      |
| Use SkSLs from MotoG4      | 31.346 | 20.232      |
| Use SkSLs from OnePlus7Pro | 33.103 | 22.638      |

See more detailed experiment result in  [shader_test.zip](https://github.com/flutter/flutter/files/4423186/shader_test.zip)

**Comment by liyuqian** [comment]
I did another experiment of using Android emulators to generate SkSLs, and use them to warm up Moto G4.

34 out of 35 SkSLs share the same keys (filenames). 24 out of 35 SkSLs share the exact same content. Below is the worst frame comparison between emulator SkSL and actual MotoG4 SkSL:

| Worst frame time (ms)      | MotoG4 | 
|----------------------------|--------|
| No cache                  | 94.535 |
| Use SkSLs from MotoG4      | 31.346 |
| Use SkSLs from OnePlus7Pro | 33.103 |
| Use SkSLs from emulator              | 42.981 |

So it's clearly worse than SkSLs from OnePlus7Pro or Moto G4, but the speedup (over no cache) is still quite significant. This suggests that if there are proper emulators on the host machine, we can generate a decent set of SkSLs without using actual mobile devices.

CC @brianosman

**Comment by liyuqian** [comment]
Another experiment was done to use iPhone 6s to generate SkSLs, and use them on Moto G4. (We still have to write some code to use let iPhone 6s use SkSLs from Moto G4, so such experiment will be conducted later.)

iPhone6s generated 37 SkSLs. About 7 of them have have different keys/filenames than Moto G4 generated SkLSs. 10 of the SkSLs have different content (including the 7 SkSLs that have different keys). The worst frame time from MotoG4 is as below:

| Worst frame time (ms)      | MotoG4 | 
|----------------------------|--------|
| No cache                  | 94.535 |
| Use SkSLs from MotoG4      | 31.346 |
| Use SkSLs from OnePlus7Pro | 33.103 |
| Use SkSLs from emulator              | 42.981 |
| Use SkSLs from iPhone6s              | 38.0 |

**Comment by liyuqian** [comment]
@brianosman : considering the surprisingly good set of SkSLs generated from the emulator without any work, do you think it would be relative easy to do the following?
1. Record the GPU/OpenGL profile of a real device
2. Let Skia simulate that profile during the emulator run to emit the exact same set of SkSLs that a real device would generate.

If so, would it make "SkSL generation on the host machine" possible with much less work than the previous estimation of one quarter?

CC @kf6gpe

**Comment by brianosman** [comment]
If the emulator itself were able to mock a specific device, this would help. Unfortunately, the biggest problem is that mobile GPUs tend to support a collection of extensions that are not supported on desktop, and (as far as I know) the emulator does nothing to correct that. (So the emulator's GLES will never advertise or support extensions that are exclusive to mobile architectures). Skia takes advantage of those extensions, because they're performance critical when they're relevant (complex blend modes, etc.).

Working around these extension incompatibilities in this scenario would be about as difficult as the original proposal (of doing it all inside Skia), so I think the time estimate would remain similar (~ one quarter).

**Comment by kf6gpe** [comment]
cc @kf6gpe

**Comment by jonahwilliams** [comment]
@liyuqian I updated the checklist with items I believe the complete. Is there any additional work needed from tools to make this work for the June milestone?

**Comment by liyuqian** [comment]
No, that's all we need from tools. Thank you so much!

**Comment by adeolaex** [comment]
Any update on the documentation ?

**Comment by DeadlyMissile** [comment]
@liyuqian Wrong link for the One Plus device :D

**Comment by liyuqian** [comment]
Thanks @DeadlyMissile for catching it! @adeolaex : I just added the doc to https://github.com/flutter/flutter/wiki/Reduce-shader-compilation-jank-using-SkSL-warm-up

**Comment by adeolaex** [comment]
@liyuqian Maybe an extra clarification for the iOS building command, on number 4 of the documentation.

**Comment by acoutts** [comment]
@liyuqian thanks for your work on this! One feedback: I was just having issues building my app and the xcode output was very verbose but I couldn't find the issue, and in the end it was because I had an invalid path for `bundle-sksl-path` but I never saw that anywhere in the build output (or it was very hard to find buried).

**Comment by liyuqian** [comment]
Thanks @acoutts for the feedback! CC @jonahwilliams to see if there's a quick tweak in the tool that can help the error message of an invalid `bundle-sksl-path` standout. Also CC @InMatrix who recently conduct some research in error messages and might be familiar on how to improve this.

**Comment by jonahwilliams** [comment]
Filled https://github.com/flutter/flutter/issues/61772 for the validation

**Comment by adeolaex** [comment]
I get this error `Oops; flutter has exited unexpectedly: "Invalid argument(s): Cannot find executable for /Users/adeola/engine/src/out/ios_release/clang_x64/gen_snapshot_arm64.".` when trying to run `flutter run` with a custom flutter engine. 

Any help as to what this means ?

**Comment by liyuqian** [comment]
@adeolaex : the issue looks the same as https://github.com/flutter/flutter/issues/38933. There's a workaround there.

**Comment by acoutts** [comment]
@liyuqian how can we run a driver test using a `flutter_01.sksl.json` sksl?

**Comment by adeolaex** [comment]
@liyuqian 
Temporary workaround is to symlink `gen_snapshot_armv7` or `gen_snapshot_arm64` to `gen_snapshot` while developing locally.

_Originally posted by @chinmaygarde in https://github.com/flutter/flutter/issues/38933#issuecomment-523688766_

That's the workaround and I'm  aware that I could use `ln -s` to create a symlink, but where can I find the `gen_snapshot_armv7` or `gen_snapshot_arm64`.

**Comment by liyuqian** [comment]
@acoutts : there's no built-in SkSL support for `flutter drive` yet so you'll have to first `flutter build`, `flutter run`, and finally `flutter drive` on a running app by giving the observatoryUri. See the following piece of code and the implementation of `_buildAndRun()` and `internalRun(existingApp: ...)` in the same file. (The `internalRun` is the `flutter drive` step).

https://github.com/flutter/flutter/blob/9e658bc3d5fa9ce3add5e90ca995b762e5135781/dev/devicelab/lib/tasks/perf_tests.dart#L459-L463

**Comment by liyuqian** [comment]
@adeolaex : the `gen_snapshot` should be in a place like `out/ios_profile/clang_x64/gen_snapshot`.

**Comment by acoutts** [comment]
> @acoutts : there's no built-in SkSL support for `flutter drive` yet so you'll have to first `flutter build`, `flutter run`, and finally `flutter drive` on a running app by giving the observatoryUri. See the following piece of code and the implementation of `_buildAndRun()` and `internalRun(existingApp: ...)` in the same file. (The `internalRun` is the `flutter drive` step).
> 
> https://github.com/flutter/flutter/blob/9e658bc3d5fa9ce3add5e90ca995b762e5135781/dev/devicelab/lib/tasks/perf_tests.dart#L459-L463

I see. One problem I have with this approach is that the most severe jank in my app is on the initial startup / run, so if I cannot run a driver test from the very start from a fresh install (with --bundle-sksl-path provided) then I cannot consistently baseline / compare the two runs.

**Comment by liyuqian** [comment]
@acoutts : in order to benchmark your startup animation performance, can you do something like the following in your `test_driver/app.dart` and `test_driver/app_test.dart`?

In `test_driver/app.dart`
```dart
void main() {
  enableFlutterDriverExtension();
  // Normally, we'd have `runApp(TheActualAppWidget())` here.
  // To test the startup animation, we'll defer loading `TheActualAppWidget` with a button.
  runApp(
      ... // Scaffold widgets
      child: RaisedButton(
        onPressed: () { 
            ... // switch to TheActualAppWidget() 
        },
        child: Text('Start test'),
      ),
  );
}
```

In `test_driver/app_test.dart`
```dart
void main() {
    ...
    final Timeline timeline = await driver.traceAction(() async {
        driver.tap(...); // tap the "Start test" button.
        ... // wait until the startup animation is done.
    });
    ...
}

```

**Comment by acoutts** [comment]
Excellent! I was able to do it in my case with a StatefulBuilder.

`test_driver/app.dart`:
```dart
void main() async {
  // This line enables the extension.
  enableFlutterDriverExtension();

  /// Make sure this to tell the app it's a driver test
  Env().isDriverTest = true;

  /// Make sure we're on sandbox
  Env().sandboxEnv = SandboxEnv.dev;

  /// Start the bottlepay app
  return app.main();
}
```

`main.dart`:
```dart
/// Now start the app
void main() {
  var appStarted = !Env().isDriverTest;
  
  runApp(
    StatefulBuilder(
      builder: (context, setState) {
        if (appStarted) {
          // actual app here
        } else {
          return Directionality(
            textDirection: ui.TextDirection.ltr,
            child: Center(
              child: RaisedButton(
                child: Text('Start app'),
                onPressed: () {
                  setState(() => appStarted = true);
                },
              ),
            ),
          );
        }
      },
    ),
  );
}
```

`test_driver/app_test.dart`:
```dart
test('initial_app_opening', () async {
  await driver.clearTimeline();
  final timeline = await driver.traceAction(() async {
    /// Tap start button
    await driver.tap(find.text('Start app'));
    await driver.waitFor(find.text('SIGNUP/LOGIN'));
  });

  /// Convert the Timeline into a TimelineSummary that's easier to read and
  /// understand.
  final summary = TimelineSummary.summarize(timeline);

  /// Then, save the summary to disk.
  await summary.writeSummaryToFile('initial_app_opening', pretty: true);

  /// Optionally, write the entire timeline to disk in a json format. This
  /// file can be opened in the Chrome browser's tracing tools found by
  /// navigating to chrome://tracing.
  await summary.writeTimelineToFile('initial_app_opening', pretty: true);
}, tags: 'driver');
```
I will update the thread here with my new test findings: https://github.com/flutter/flutter/issues/61045

**Comment by acoutts** [comment]
Ugh, I am not having annoying environmental issues. An hour ago I was running the driver tests no problem. Now it is hanging indefinitely on `await driver.waitUntilFirstFrameRasterized();` after the isolates are unpaused, and nothing ever runs / starts.

Are the driver tests always this flaky? :-/

**Comment by liyuqian** [comment]
@acoutts : yes, we're aware of the flakiness of driver tests. @CareF is working on https://github.com/orgs/flutter/projects/8#card-41373712 to help alleviate the problem.

**Comment by scutlight** [comment]
@liyuqian : For add-to-app scenario, precompile SkSL shaders happens when the first time the engine start rendering to a surface, that is also the first frame rendering occurs. As a result, it might postpone the first frame task on the raster thread. Is it possible to make the warmup more advance? Seems it can only after `eglMakeCurrent`?

**Comment by liyuqian** [comment]
It does need a valid GLContext. Currently it should be warming up SkSL shaders as soon as the GPUSurfaceGL is set up before the first frame is requested. In order to reduce add-to-app latency, one should be able to [pre-warm][1] the engine before user taps a button to trigger the add-to-app Flutter view. If `GLContextMakeCurrent` has to be called before warmup, maybe the app can either create an offscreen context, or an invisible on-screen context (e.g., fully clipped out, fully transparent, etc.), during the engine pre-warm-up?

CC @xster and @gaaclarke who I believe were previously involved in iOS add-to-app startup pre-warm discussions.

[1]: https://flutter.dev/docs/development/add-to-app/performance#memory-and-latency

**Comment by LiangLuDev** [comment]
how to 'flutter_01.sksl.json' build ios testflight？

**Comment by IvanSyabro** [comment]
Hello, I would like to know if there is any way to use [SKSL warmup](https://flutter.dev/docs/perf/rendering/shader#how-to-use-sksl-warmup) when flutter is added as module to main android project, as described here https://flutter.dev/docs/development/add-to-app/android/project-setup. Would be grateful if someone provided a hint.

**Comment by erf** [comment]
@liyuqian when i try to run 

```flutter run --profile --cache-sksl --purge-persistent-cache```

i only get 

```Error running application on ...```

I've tried to run this on both **Dev** and **Beta**.

```
flutter --version

Flutter 2.3.0-24.1.pre • channel beta • https://github.com/flutter/flutter.gitFramework • revision 615957513e (5 days ago) • 2021-06-25 09:24:44 -0700
Engine • revision 9d517f475b
Tools • Dart 2.14.0 (build 2.14.0-188.5.beta)
```
